### ISSUE TITLE
UTF-8 encoding detection fails on PHP 8.1

### ISSUE THREAD

**Issue by come-nc** [sabre-io/http#181]
Hello

Someone openned an issue at Nextcloud signaling that creating a folder with special character broke when using PHP 8.1.
It appears that this comes from mb_detect_encoding changes in PHP 8.1 that end up detecting ISO-8859-1 instead of UTF-8

The code handling this is the same as the one in you decodePathSegment at https://github.com/sabre-io/http/blob/master/lib/functions.php#L404
You can see the behavior here: https://3v4l.org/IT9dJ

It feels we should not be trying to detect encoding here but maybe assume UTF-8, or only check with mb_check_encoding (which does return true for UTF-8).
What is the reason for this test, are some dav client sending ISO-8859-1? Is that allowed by webdav protocol?

Would this be ok instead:
```php
/**
 * Decodes a url-encoded path segment.
 */
function decodePathSegment(string $path): string
{
    $path = rawurldecode($path);

    if (!mb_check_encoding($path, 'UTF-8') && mb_check_encoding($path, 'ISO-8859-1')) {
        $path = utf8_encode($path);
    }

    return $path;
}
```

**Comment by staabm** [comment]
thx for the report

> Someone openned an issue at Nextcloud signaling that creating a folder with special character broke when using PHP 8.1.
> It appears that this comes from mb_detect_encoding changes in PHP 8.1 that end up detecting ISO-8859-1 instead of UTF-8

did you check whether its a intended change in php-src or maybe even a php 8.1 only bug?
I would suggest opening a php-src issue and see whether it is expected.

**Comment by come-nc** [comment]
Opened https://github.com/php/php-src/issues/8279

**Comment by come-nc** [comment]
Note that there is the same problem in https://github.com/sabre-io/dav/blob/master/lib/DAV/StringUtil.php#L78 and https://github.com/sabre-io/vobject/blob/master/lib/StringUtil.php#L41

**Comment by staabm** [comment]
> See the PHP dev answer, this change is on purpose. Can you clarify why encoding detection is needed in this case? Is the webdav protocol accepting any encoding or is it because some clients are misbehaving?

I can't say anything about the protocol, but technically the detection is needed because `utf8_encode` only works on `ISO-8859-1`

**Comment by staabm** [comment]
> But how come we do not know the encoding of what we are receiving in the first place?

I don't know.


> Would you be ok with switching to `mb_check_encoding` as proposed in my first post?

Yep, after the discussion it seems the way forward.

**Comment by DeepDiver1975** [comment]
> But how come we do not know the encoding of what we are receiving in the first place?

I can only guess here .... sabre dav has a long history and a lot of bug fixing went into buggy clients sending essentially :hankey: .
I guess sabre is not trusting any client and performs additional checks/verification/validations ...

**Comment by staabm** [comment]
btw: I tracked back the history when this was added and found the commit https://github.com/sabre-io/http/commit/885ada52bca251f7c0dba7bf05d9848afb7e6385#diff-c8569a19ab0202cf69472c1bc9c1945632a552d660283e3c8354408803a12972R79

**Comment by ambraspace** [comment]
@come-nc thanks for investigating this issue.

I reported this to Nextcloud repository since I noticed the problem, not only with file names but also with contact names. Since (probably) more then 90% names in Slavic languages contain one of šđčćž it's quite a problem.

After reading the discussion at https://github.com/php/php-src/issues/8279 I'm even more worried. It seems that sometimes PHP will guess UTF-8 and sometimes it will guess ISO-8859-1 for the same application. It seems very awkward. Shouldn't the encoding be detected from the HTTP request header or something?

Unless there's is a way to specify the encoding within WebDAV protocol, I strongly suggest assuming UTF-8 by default.
